crates/vm/levm/src/precompiles.rs

```diff
@@ -678,12 +680,211 @@ fn update_pairing_result(
+// four words indexed by "a", "b", "c", and "d" in the working vector
+// v[0..15].  The full modified vector is returned.
+// Based on https://datatracker.ietf.org/doc/html/rfc7693#section-3.1
+#[allow(clippy::indexing_slicing)]
+fn g(v: [u64; 16], a: usize, b: usize, c: usize, d: usize, x: u64, y: u64) -> [u64; 16] {
+    let mut ret = v;
+    ret[a] = v[a].wrapping_add(v[b]).wrapping_add(x);
+    ret[d] = (ret[d] ^ ret[a]).rotate_right(R1);
+    ret[c] = ret[c].wrapping_add(ret[d]);
+    ret[b] = (ret[b] ^ ret[c]).rotate_right(R2);
+    ret[a] = ret[a].wrapping_add(ret[b]).wrapping_add(y);
+    ret[d] = (ret[d] ^ ret[a]).rotate_right(R3);
+    ret[c] = ret[c].wrapping_add(ret[d]);
+    ret[b] = (ret[b] ^ ret[c]).rotate_right(R4);
+
+    ret
+}
+
+// Perform the permutations on the work vector
+#[allow(clippy::indexing_slicing)]
+fn word_permutation(rounds_to_permute: usize, v: [u64; 16], m: &[u64; 16]) -> [u64; 16] {
+    let mut ret = v;
+
+    for i in 0..rounds_to_permute {
+        // Message word selection permutation for each round.
+
+        let s: &[usize; 16] = &SIGMA[i % 10];
+
+        ret = g(ret, 0, 4, 8, 12, m[s[0]], m[s[1]]);
+        ret = g(ret, 1, 5, 9, 13, m[s[2]], m[s[3]]);
+        ret = g(ret, 2, 6, 10, 14, m[s[4]], m[s[5]]);
+        ret = g(ret, 3, 7, 11, 15, m[s[6]], m[s[7]]);
+
+        ret = g(ret, 0, 5, 10, 15, m[s[8]], m[s[9]]);
+        ret = g(ret, 1, 6, 11, 12, m[s[10]], m[s[11]]);
+        ret = g(ret, 2, 7, 8, 13, m[s[12]], m[s[13]]);
+        ret = g(ret, 3, 4, 9, 14, m[s[14]], m[s[15]]);
+    }
+
+    ret
+}
+
+// Based on https://datatracker.ietf.org/doc/html/rfc7693#section-3.2
+fn blake2f_compress_f(
+    rounds: usize,
+    h: [u64; 8],
+    m: &[u64; 16],
+    t: &[u64; 2],
+    f: bool,
+) -> Result<[u64; 8], InternalError> {
+    // Initialize local work vector v[0..15], takes first half from state and second half from IV.
+    let mut v: [u64; 16] = [0; 16];
+    v[0..8].copy_from_slice(&h);
+    v[8..16].copy_from_slice(&IV);
+
+    v[12] ^= t[0]; // Low word of the offset
+    v[13] ^= t[1]; // High word of the offset
+
+    // If final block flag is true, invert all bits
+    if f {
+        v[14] = !v[14];
+    }
+
+    v = word_permutation(rounds, v, m);
+
+    let mut output = [0; 8];
+
+    // XOR the two halves, put the results in the output slice
+    for (i, pos) in output.iter_mut().enumerate() {
+        *pos = h.get(i).ok_or(InternalError::SlicingError)?
+            ^ v.get(i).ok_or(InternalError::SlicingError)?
+            ^ v.get(i.overflowing_add(8).0)
+                .ok_or(InternalError::SlicingError)?;
+    }
+
+    Ok(output)
+}
+
+pub fn blake2f(
+    calldata: &Bytes,
+    gas_for_call: u64,
+    consumed_gas: &mut u64,
 ) -> Result<Bytes, VMError> {
-    Ok(Bytes::new())
+    if calldata.len() != 213 {
+        return Err(VMError::PrecompileError(PrecompileError::ParsingInputError));
+    }
+
+    let rounds: U256 = calldata
```

Comment:
why `U256` for a 4 byte value?

Reply:
It's an intermediate size to then move on to a usize, I couldn't find a prettier way to do it.